source/beamline/BioXAS/BioXASValvesControl.cpp

```diff
@@ -0,0 +1,421 @@
+#include "BioXASValvesControl.h"
+#include "actions3/AMListAction3.h"
+#include "actions3/AMActionSupport.h"
+
+BioXASValvesControl::BioXASValvesControl(const QString &name, QObject *parent) :
+	AMPseudoMotorControl(name, "", parent)
+{
+	// Initialize inherited variables.
+
+	value_ = None;
+	setpoint_ = None;
+	minimumValue_ = None;
+	maximumValue_ = Closed;
+
+	setEnumStates(QStringList() << valueToString(None) << valueToString(Open) << valueToString(Closed));
+	setMoveEnumStates(QStringList() << valueToString(Open));
+
+	setContextKnownDescription("ValvesControl");
+
+	// Create member variables.
+
+	frontEndValveSet_ = 0;
+	sideValveSet_ = 0;
+	mainValveSet_ = 0;
+	imagingValveSet_ = 0;
+
+	// Current settings.
+
+	updateStates();
+}
+
+BioXASValvesControl::~BioXASValvesControl()
+{
+
+}
+
+bool BioXASValvesControl::canMeasure() const
+{
+	return false;
+}
+
+bool BioXASValvesControl::canMove() const
+{
+	return false;
+}
+
+bool BioXASValvesControl::validValue(double value) const
+{
+	bool result = false;
+
+	switch (int(value)) {
+	case None:
+		result = true;
+		break;
+	case Open:
+		result = true;
+		break;
+	case Closed:
+		result = true;
+		break;
+	default:
+		break;
+	}
+
+	return result;
+}
+
+bool BioXASValvesControl::validSetpoint(double value) const
+{
+	bool result = false;
+
+	// Only want to be able to use this control to view current
+	// state, and to open valves (not close them).
+
+	if (value == BioXASValvesControl::Open)
+		result = true;
+
+	return result;
+}
+
+AMControlSet* BioXASValvesControl::valveSet(BioXAS::Beamline beamline)
+{
+	AMControlSet *result = 0;
+
+	switch (beamline) {
+	case BioXAS::FrontEnd:
+		result = frontEndValveSet_;
+		break;
+	case BioXAS::Side:
+		result = sideValveSet_;
+		break;
+	case BioXAS::Main:
+		result = mainValveSet_;
+		break;
+	case BioXAS::Imaging:
+		result = imagingValveSet_;
+		break;
+	default:
+		break;
+	}
+
+	return result;
+}
+
+bool BioXASValvesControl::isOpen() const
```

Comment:
This valves control is going to be used in the beam on/off controls for BioXAS, and the requirement for 'beam on' for the front end is that all valves on Side, Main, and Imaging are all open. Any other state would be be indicative of 'beam off'. I was thinking about this control from that perspective, though I see it can certainly be made to be more flexible. Is your recommendation to: 
- add additional 'Between' state
- consider only two states: Open, NotOpen